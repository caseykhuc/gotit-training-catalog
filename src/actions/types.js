export const userTypes = {
  // user
  REGISTER_USER: 'REGISTER_USER',
  REGISTER_USER_REQUEST: 'REGISTER_USER_REQUEST',
  REGISTER_USER_SUCCESS: 'REGISTER_USER_SUCCESS',
  REGISTER_USER_FAILURE: 'REGISTER_USER_FAILURE',

  LOGIN_USER: 'LOGIN_USER',
  LOGIN_USER_REQUEST: 'LOGIN_USER_REQUEST',
  LOGIN_USER_SUCCESS: 'LOGIN_USER_SUCCESS',
  LOGIN_USER_FAILURE: 'LOGIN_USER_FAILURE',

  LOGOUT_USER: 'LOG_OUT',
};

export const itemTypes = {
  // items
  FETCH_ITEMS: 'FETCH_ITEMS',
  FETCH_ITEMS_REQUEST: 'FETCH_ITEMS_REQUEST',
  FETCH_ITEMS_SUCCESS: 'FETCH_ITEMS_SUCCESS',
  FETCH_ITEMS_FAILURE: 'FETCH_ITEMS_FAILURE',

  FETCH_ITEM: 'FETCH_ITEM',
  FETCH_ITEM_REQUEST: 'FETCH_ITEM_REQUEST',
  FETCH_ITEM_SUCCESS: 'FETCH_ITEM_SUCCESS',
  FETCH_ITEM_FAILURE: 'FETCH_ITEM_FAILURE',

  EDIT_ITEM: 'EDIT_ITEM',
  EDIT_ITEM_REQUEST: 'EDIT_ITEM_REQUEST',
  EDIT_ITEM_SUCCESS: 'EDIT_ITEM_SUCCESS',
  EDIT_ITEM_FAILURE: 'EDIT_ITEM_FAILURE',

  DELETE_ITEM: 'DELETE_ITEM',
  DELETE_ITEM_REQUEST: 'DELETE_ITEM_REQUEST',
  DELETE_ITEM_SUCCESS: 'DELETE_ITEM_SUCCESS',
  DELETE_ITEM_FAILURE: 'DELETE_ITEM_FAILURE',

  ADD_ITEM: 'ADD_ITEM',
  ADD_ITEM_REQUEST: 'ADD_ITEM_REQUEST',
  ADD_ITEM_SUCCESS: 'ADD_ITEM_SUCCESS',
  ADD_ITEM_FAILURE: 'ADD_ITEM_FAILURE',
};

export const categoryTypes = {
  // category
  FETCH_CATEGORY: 'FETCH_CATEGORY',
  FETCH_CATEGORY_REQUEST: 'FETCH_CATEGORY_REQUEST',
  FETCH_CATEGORY_SUCCESS: 'FETCH_CATEGORY_SUCCESS',
  FETCH_CATEGORY_FAILURE: 'FETCH_CATEGORY_FAILURE',
};
