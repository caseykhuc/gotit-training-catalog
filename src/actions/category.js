import { categoryTypes } from '../constants/actionTypes';
import * as api from '../api';

// eslint-disable-next-line import/prefer-default-export
export const fetchCategory = () => ({
  type: categoryTypes.FETCH_CATEGORY,
  promise: api.fetchCategories(),
})
