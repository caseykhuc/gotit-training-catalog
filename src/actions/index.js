import actionTypes from '../const/types';
import { request } from '../api';

/* export const fetchItem = () => ({
  type: actionTypes.FETCH_ITEM,
  promise: request('/item', { method: 'GET' }),
}); */

/* body = {
  'username': 'username',
  'email': 'email',
  'name': 'name',
  'password': 'password'
}; */
export const registerUser = (body) => ({
  type: actionTypes.SIGNUP_USER,
  promise: request('/user', {
    method: 'POST',
    body: JSON.stringify(body),
    headers: { 'Content-Type': 'application/json' },
  }),
});

/* body = {
  'username': 'username',
  'email': 'email',
  'name': 'name',
  'password': 'password'
}; */
export const loginUser = (body) => ({
  type: actionTypes.LOGIN_USER,
  promise: request('/user', {
    method: 'POST',
    body: JSON.stringify(body),
    headers: { 'Content-Type': 'application/json' },
  }),
});

export const logoutUser = () => ({
  //sync
});

export const fetchCategory = (offset = 0, limit = 10) => ({
  type: actionTypes.FETCH_CATEGORY,
  promise: request(`/categories/?offset=${offset}&limit=${limit}`, {
    method: 'GET',
  }),
});

export const fetchItems = (categoryId, offset = 0, limit = 10) => ({
  //
});

export const addItem = () => {};
export const editItem = () => {};

export const deleteItemAndRefetch = () => (dispatch) => {};


