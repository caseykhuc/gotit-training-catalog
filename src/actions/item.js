import { itemTypes } from '../constants/actionTypes';
import * as api from '../api';

/* export const fetchItem = () => ({
  type: actionTypes.FETCH_ITEM,
  promise: request('/item', { method: 'GET' }),
}); */
export const fetchItems = (categoryId, page = 0) => ({
  type: itemTypes.FETCH_ITEMS,
  promise: api.fetchItems(categoryId, page),
})

export const addItem = () => { };
export const editItem = () => { };

export const deleteItemAndRefetch = () => (dispatch) => { };
