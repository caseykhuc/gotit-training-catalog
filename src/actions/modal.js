import { modalTypes } from 'constants/actionTypes';

export const showModal = (modalKey, props = {}) => ({
  type: modalTypes.SHOW_MODAL,
  payload: {
    modalKey,
    props,
  },
});

export const hideModal = () => ({
  type: modalTypes.HIDE_MODAL,
});
