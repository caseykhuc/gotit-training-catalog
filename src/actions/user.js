import { userTypes } from '../constants/actionTypes';
import { request } from '../api';

/* body = {
  'username': 'username',
  'email': 'email',
  'name': 'name',
  'password': 'password'
}; */
export const registerUser = (username, password) => ({
  type: userTypes.REGISTER_USER,
  payload: { username, password },
  /* promise: request('/user', {
    method: 'POST',
    body: JSON.stringify(body),
    headers: { 'Content-Type': 'application/json' },
  }), */
});

/* body = {
  'username': 'username',
  'email': 'email',
  'name': 'name',
  'password': 'password'
}; */
export const signinUser = (body) => ({
  type: userTypes.LOGIN_USER,
  promise: request('/user', {
    method: 'POST',
    body: JSON.stringify(body),
    headers: { 'Content-Type': 'application/json' },
  }),
});

export const signoutUser = () => ({
  // sync
});
