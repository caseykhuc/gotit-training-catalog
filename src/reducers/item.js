import { combineReducers } from 'redux';
import { itemTypes } from 'constants/actionTypes';
import { arrayToObject } from 'utils';

/** store items in the current page
 * or about a single item itself
*/
const byId = (state = {}, action) => {
  switch (action.type) {
    case itemTypes.FETCH_ITEMS_SUCCESS:
      return arrayToObject(action.payload.items);
    case itemTypes.FETCH_ITEM_SUCCESS:
      return { ...state, [action.payload.id]: action.payload };
    case itemTypes.EDIT_ITEM_SUCCESS:
      return { ...state, [action.payload.id]: action.payload };
    /** do not modify state here because
     * deleted/added item required refetch items
    case itemTypes.DELETE_ITEM_SUCCESS:
    case itemTypes.ADD_ITEM_SUCCESS:
      return state;
      */
    default:
      return state;
  }
};

const isLoading = (state = false, action) => {
  switch (action.type) {
    case itemTypes.ADD_ITEM_REQUEST:
    case itemTypes.DELETE_ITEM_REQUEST:
    case itemTypes.EDIT_ITEM_REQUEST:
    case itemTypes.FETCH_ITEMS_REQUEST:
    case itemTypes.FETCH_ITEM_REQUEST:
      return true;

    case itemTypes.ADD_ITEM_SUCCESS:
    case itemTypes.DELETE_ITEM_SUCCESS:
    case itemTypes.EDIT_ITEM_SUCCESS:
    case itemTypes.FETCH_ITEMS_SUCCESS:
    case itemTypes.FETCH_ITEM_SUCCESS:
    case itemTypes.ADD_ITEM_FAILURE:
    case itemTypes.DELETE_ITEM_FAILURE:
    case itemTypes.EDIT_ITEM_FAILURE:
    case itemTypes.FETCH_ITEMS_FAILURE:
    case itemTypes.FETCH_ITEM_FAILURE:
      return false;

    default:
      return state;
  }
};

const error = (state = '', action) => {
  switch (action.type) {
    case itemTypes.ADD_ITEM_FAILURE:
    case itemTypes.EDIT_ITEM_FAILURE:
    case itemTypes.DELETE_ITEM_FAILURE:
    case itemTypes.FETCH_ITEMS_FAILURE:
    case itemTypes.FETCH_ITEM_FAILURE:
      return action.payload;
    case itemTypes.ADD_ITEM_REQUEST:
    case itemTypes.DELETE_ITEM_REQUEST:
    case itemTypes.EDIT_ITEM_REQUEST:
    case itemTypes.FETCH_ITEMS_REQUEST:
    case itemTypes.FETCH_ITEM_REQUEST:
    case itemTypes.ADD_ITEM_SUCCESS:
    case itemTypes.EDIT_ITEM_SUCCESS:
    case itemTypes.FETCH_ITEMS_SUCCESS:
    case itemTypes.FETCH_ITEM_SUCCESS:
      return '';
    default:
      return state;
  }
};

// totalItems in current category
const totalItems = (state = 0, action) => {
  switch (action.type) {
    case itemTypes.FETCH_ITEMS_SUCCESS:
      return action.payload.totalItems;
    case itemTypes.ADD_ITEM_SUCCESS:
      return state + 1;
    case itemTypes.DELETE_ITEM_SUCCESS:
      return state - 1;
    default:
      return state;
  }
};

const page = (state = 0, action) => {
  switch (action.type) {
    case itemTypes.FETCH_ITEMS_SUCCESS:
      return action.payload.page;
    default:
      return state;
  }
}

export default combineReducers({
  byId, totalItems, isLoading, error, page,
});

export const getItem = (state, id) => state.byId[id];

export const getItems = (state) => (state.byId
  ? Object.keys(state.byId).map((id) => state.byId[id])
  : {});
