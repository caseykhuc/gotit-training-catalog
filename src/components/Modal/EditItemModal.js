import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import validator from 'validator';

import BaseFormModal from 'components/Base/BaseFormModal';
import { editItem } from 'actions/item';
import * as selector from 'reducers';
import * as errorMessage from 'constants/inputError';

export const validate = ({
  name, price,
}) => {
  const priceString = String(price);
  const inputError = {};
  if (name && name.length < 5) inputError.name = errorMessage.name.tooShort;
  if (priceString && !validator.isNumeric(priceString)) {
    inputError.price = errorMessage.price.isNotNumeric;
  }
  return inputError;
};

export const EditItemModal = ({
  editItem, currentValue = {},
}) => {
  const initialState = {
    inputValue: currentValue,
    inputError: {
    },
    requestError: '',
  };

  const fields = [
    { name: 'name', type: 'text' },
    { name: 'description', type: 'textarea' },
    { name: 'price', type: 'text' },
  ];

  return (
    <BaseFormModal
      title="EDIT ITEM"
      fields={fields}
      initialState={initialState}
      onAction={editItem}
      validate={validate}
    />
  )
}

EditItemModal.propTypes = {
  editItem: PropTypes.func.isRequired,
  currentValue: PropTypes.shape({
    name: PropTypes.string,
    description: PropTypes.string,
    price: PropTypes.number,
  }),
}

const mapStateToProps = (state) => ({
  categories: selector.getCategories(state),
})

export const mapDispatchToProps = (dispatch, { categoryId, itemId }) => ({
  editItem: (body) => dispatch(editItem(categoryId, itemId)(body)),
})

export default connect(mapStateToProps, mapDispatchToProps)(EditItemModal);
