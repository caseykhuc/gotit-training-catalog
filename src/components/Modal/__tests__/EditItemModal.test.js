import React from 'react';
import { shallow } from 'enzyme';
import { editItem } from 'actions/item';
import { EditItemModal, validate, mapDispatchToProps } from '../EditItemModal';

describe('components/Common/EditItemModal', () => {
  let props;
  let wrapper;

  const update = () => {
    wrapper.update();
  };
  const setup = () => {
    wrapper = shallow(<EditItemModal {...props} />);
    update();
  };
  beforeEach(() => {
    props = {
      editItem: jest.fn().mockResolvedValue({}),
      currentValue: {
        name: 'Sample',
        description: '',
        price: 100,
      },
    }
  });
  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should set default value for currentValue', () => {
    props.currentValue = undefined;
    setup();
    expect(wrapper.props().initialState.inputValue).toEqual({});
  })
})

describe('components/Modal/EditItemModal (validate)', () => {
  let input;
  beforeEach(() => {
    input = {
      name: 'Name',
      price: '',
    }
  });
  it('should render error for name shorter then 5 characters', () => {
    expect(validate(input).name).toBeTruthy();
    input.name = 'Longer Name';
    expect(validate(input).name).toBeFalsy();
  });
  it('should render error for not numeric price', () => {
    expect(validate(input).price).toBeFalsy();
    input.price = 'abc';
    expect(validate(input).price).toBeTruthy();
  });
})

describe('components/Modal/EditItemModal (mapDispatchToProps)', () => {
  const dispatch = jest.fn();
  it('should return dispatched action', () => {
    const dispatched = mapDispatchToProps(dispatch,
      { categoryId: 1, itemId: 2 }).editItem;
    dispatched();
    expect(dispatch).toHaveBeenCalledWith(editItem(1, 2)())
  });
})
