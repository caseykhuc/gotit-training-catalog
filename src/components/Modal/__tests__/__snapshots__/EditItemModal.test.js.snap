// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Common/EditItemModal should render correctly 1`] = `ShallowWrapper {}`;
